components/nexusoperations/link_converter.go

```
@@ -0,0 +1,130 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package nexusoperations
+
+import (
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/nexus-rpc/sdk-go/nexus"
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/api/enums/v1"
+)
+
+func ConvertLinkWorkflowEventToNexusLink(we *commonpb.Link_WorkflowEvent) (nexus.Link, error) {
+	u, err := url.Parse(fmt.Sprintf(
+		"temporal:///namespaces/%s/workflows/%s/%s/history",
+		we.GetNamespace(),
+		we.GetWorkflowId(),
+		we.GetRunId(),
+	))
+	if err != nil {
+		return nexus.Link{}, err
+	}
+
+	linkType := ""
+	switch ref := we.GetReference().(type) {
+	case *commonpb.Link_WorkflowEvent_EventRef:
+		u.RawQuery = convertLinkWorkflowEventEventReferenceToURLQuery(ref.EventRef)
+		linkType = string(ref.EventRef.ProtoReflect().Descriptor().FullName())
+	}
+	return nexus.Link{URL: u, Type: linkType}, nil
+}
+
+func ConvertNexusLinkToLinkWorkflowEvent(link nexus.Link) (*commonpb.Link_WorkflowEvent, error) {
+	we := &commonpb.Link_WorkflowEvent{}
+	if !strings.HasPrefix(link.Type, string(we.ProtoReflect().Descriptor().FullName())+".") {
+		return nil, fmt.Errorf(
+			"cannot parse link type %q to %q",
+			link.Type,
+			we.ProtoReflect().Descriptor().FullName(),
+		)
+	}
+
+	pathParts := strings.Split(link.URL.Path, "/")
+	if len(pathParts) != 7 {
+		return nil, fmt.Errorf("failed to parse link to Link_WorkflowEvent")
+	}
+	if pathParts[0] != "" || pathParts[1] != "namespaces" || pathParts[3] != "workflows" || pathParts[6] != "history" {
+		return nil, fmt.Errorf("failed to parse link to Link_WorkflowEvent")
+	}
+	we.Namespace = pathParts[2]
+	we.WorkflowId = pathParts[4]
+	we.RunId = pathParts[5]
+	switch link.Type {
+	case string((&commonpb.Link_WorkflowEvent_EventReference{}).ProtoReflect().Descriptor().FullName()):
+		eventRef, err := convertURLQueryToLinkWorkflowEventEventReference(link.URL.RawQuery)
```

Why parse the query? You have it parsed already.

---

What do you mean? This is parsing the event reference.

---

I mean that you pass `RawQuery` instead of `Query`

---

Oh, well. `url.Parse` doesn't really check if the query string is valid. It silently ignore keys/values that cannot be parsed.

---

That's good, we can ignore those too.